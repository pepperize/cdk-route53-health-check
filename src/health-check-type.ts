/**
 * The type of health check to create.
 *
 * @internal
 */
export enum HealthCheckType {
  HTTP = "HTTP",
  HTTPS = "HTTPS",
  HTTP_STR_MATCH = "HTTP_STR_MATCH",
  HTTPS_STR_MATCH = "HTTPS_STR_MATCH",
  TCP = "TCP",
  CALCULATED = "CALCULATED",
  CLOUDWATCH_METRIC = "CLOUDWATCH_METRIC",
  RECOVERY_CONTROL = "RECOVERY_CONTROL",
}
