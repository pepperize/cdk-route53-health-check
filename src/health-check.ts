import { ITaggable, Resource, TagManager, TagType } from "aws-cdk-lib";
import * as cloudwatch from "aws-cdk-lib/aws-cloudwatch";

/**
 * @internal
 */
export interface HealthCheckOptions {
  /**
   * The display name of this Route53 HealthCheck
   */
  readonly healthCheckName?: string;
  /**
   * Whether to invert the status of the Route53 health check status.
   */
  readonly inverted?: boolean;
}

export interface IHealthCheck {
  readonly healthCheckId: string;
  /**
   * Return the given named metric for this HealthCheck.
   */
  metric(metricName: string, props?: cloudwatch.MetricOptions): cloudwatch.Metric;

  /**
   * Route53 health checkers report that the HealthCheck is healthy or unhealthy.
   */
  metricHealthCheckStatus(props?: cloudwatch.MetricOptions): cloudwatch.Metric;
}

/**
 * @internal
 */
export abstract class HealthCheckBase extends Resource implements IHealthCheck, ITaggable {
  public abstract readonly healthCheckId: string;

  readonly tags = new TagManager(TagType.STANDARD, "AWS::Route53::HealthCheck");

  /**
   * Return the given named metric for this HealthCheck.
   */
  metric(metricName: string, props?: cloudwatch.MetricOptions): cloudwatch.Metric {
    return new cloudwatch.Metric({
      namespace: "AWS/Route53",
      metricName,
      dimensionsMap: { HealthCheckId: this.healthCheckId },
      ...props,
    }).attachTo(this);
  }
  /**
   * Route53 health checkers report that the HealthCheck is healthy or unhealthy.
   */
  metricHealthCheckStatus(props?: cloudwatch.MetricOptions): cloudwatch.Metric {
    return this.metric("HealthCheckStatus", { statistic: cloudwatch.Statistic.MINIMUM, ...props });
  }
}

/**
 * The type of health check to create.
 */
export enum HealthCheckType {
  HTTP = "HTTP",
  HTTPS = "HTTPS",
  HTTP_STR_MATCH = "HTTP_STR_MATCH",
  HTTPS_STR_MATCH = "HTTPS_STR_MATCH",
  TCP = "TCP",
  CALCULATED = "CALCULATED",
  CLOUDWATCH_METRIC = "CLOUDWATCH_METRIC",
  RECOVERY_CONTROL = "RECOVERY_CONTROL",
}
