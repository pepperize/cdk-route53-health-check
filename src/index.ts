export * from "./alarm-health-check";
export * from "./calculated-health-check";
export * from "./endpoint-health-check";
export * from "./health-check";
