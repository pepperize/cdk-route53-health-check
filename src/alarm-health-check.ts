import { ArnFormat, Stack, Tags } from "aws-cdk-lib";
import * as cloudwatch from "aws-cdk-lib/aws-cloudwatch";
import * as route53 from "aws-cdk-lib/aws-route53";
import { Construct } from "constructs";
import { HealthCheckBase, HealthCheckOptions } from "./health-check";
import { HealthCheckType } from "./health-check-type";

export interface AlarmHealthCheckProps extends HealthCheckOptions {
  /**
   * The alarm that Route53 monitors
   */
  readonly alarm: cloudwatch.IAlarm;
  /**
   * The status to assign to the HealthCheck when CloudWatch has insufficient data about the metric
   */
  readonly insufficientDataHealthStatus?: InsufficientDataHealthStatus;
}

/**
 * Create a Route53 HealthCheck that monitors a CloudWatch Alarm.
 *
 * <b>Example</b>
 * ```typescript
 * const alarm new Alarm(stack, "Alarm", {
 *   // ...
 * });
 * new AlarmHealthCheck(stack, "HealthCheck", {
 *   alarm: alarm,
 * });
 * ```
 * @link https://docs.aws.amazon.com/de_de/AWSCloudFormation/latest/UserGuide/aws-resource-route53-healthcheck.html#aws-resource-route53-healthcheck-properties
 *
 * @resource AWS::Route53::HealthCheck
 */
export class AlarmHealthCheck extends HealthCheckBase {
  public readonly healthCheckId: string;

  constructor(scope: Construct, id: string, props: AlarmHealthCheckProps) {
    super(scope, id);

    const { alarmName, alarmArn } = props.alarm;
    // https://docs.aws.amazon.com/service-authorization/latest/reference/list_amazoncloudwatch.html
    const alarmRegion = Stack.of(props.alarm).splitArn(alarmArn, ArnFormat.COLON_RESOURCE_NAME).region;

    const insufficientDataHealthStatus =
      props.insufficientDataHealthStatus ?? InsufficientDataHealthStatus.LAST_KNOWN_STATUS;

    const resource = new route53.CfnHealthCheck(this, "Resource", {
      healthCheckConfig: {
        alarmIdentifier: {
          name: alarmName,
          region: alarmRegion!,
        },
        insufficientDataHealthStatus: insufficientDataHealthStatus,
        type: HealthCheckType.CLOUDWATCH_METRIC,
      },
      healthCheckTags: this.tags.renderedTags,
    });

    this.healthCheckId = resource.attrHealthCheckId;

    if (props.healthCheckName) {
      Tags.of(this).add("Name", props.healthCheckName);
    }
  }
}

export enum InsufficientDataHealthStatus {
  /**
   * Route53 considers the health check to be healthy.
   */
  HEALTHY = "Healthy",
  /**
   * Route53 considers the health check to be unhealthy.
   */
  UNHEALTHY = "Unhealthy",
  /**
   * Route53 uses the status of the health check from the last time that CloudWatch had sufficient data to determine the alarm state, otherwise healthy
   */
  LAST_KNOWN_STATUS = "LastKnownStatus",
}
