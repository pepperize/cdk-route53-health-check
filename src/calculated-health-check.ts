import { Annotations, Lazy, Tags } from "aws-cdk-lib";
import * as cloudwatch from "aws-cdk-lib/aws-cloudwatch";
import * as route53 from "aws-cdk-lib/aws-route53";
import { Construct } from "constructs";
import { HealthCheckBase, HealthCheckOptions, HealthCheckType, IHealthCheck } from "./health-check";

export interface CalculatedHealthCheckProps extends HealthCheckOptions {
  /**
   * The list of HealthCheck that monitors other Route53 HealthChecks.
   */
  readonly childHealthChecks?: IHealthCheck[];
  /**
   * The number of child HealthChecks that Amazon Route53 must consider healthy
   */
  readonly healthThreshold?: number;
}

/**
 * Create a Route53 HealthCheck that monitors other Route53 HealthChecks.
 *
 * <b>Example</b>
 * ```typescript
 * const healthCheck = new EndpointHealthCheck(stack, "HealthCheck", {
 *   domainName: "pepperize.com",
 * });
 * new CalculatedHealthCheck(stack, "CalculatedHealthCheck", {
 *   childHealthChecks: [healthCheck],
 * });
 * ```
 * @link https://docs.aws.amazon.com/de_de/AWSCloudFormation/latest/UserGuide/aws-resource-route53-healthcheck.html#aws-resource-route53-healthcheck-properties
 */
export class CalculatedHealthCheck extends HealthCheckBase {
  public readonly healthCheckId: string;

  private childHealthChecks: IHealthCheck[];

  constructor(scope: Construct, id: string, props: CalculatedHealthCheckProps) {
    super(scope, id);

    this.childHealthChecks = props.childHealthChecks ?? [];

    if (this.childHealthChecks.length > 256) {
      Annotations.of(this).addError("ChildHealthChecks has to be smaller than 256");
    }

    if (undefined != props.healthThreshold && props.healthThreshold > 256) {
      Annotations.of(this).addError("HealthThreshold has to be smaller than 256");
    }

    const resource = new route53.CfnHealthCheck(this, "Resource", {
      healthCheckConfig: {
        childHealthChecks: Lazy.list({
          produce: () => this.childHealthChecks.map((healthCheck) => healthCheck.healthCheckId),
        }),
        healthThreshold: props.healthThreshold,
        type: HealthCheckType.CALCULATED,
      },
      healthCheckTags: this.tags.renderedTags,
    });

    this.healthCheckId = resource.attrHealthCheckId;

    if (props.healthCheckName) {
      Tags.of(this).add("Name", props.healthCheckName);
    }
  }

  /**
   * The number of ChildHealthChecks that are healthy that Route53 is monitoring.
   *
   * Valid statistics: Average (recommended), Minimum, Maximum
   */
  public metricChildHealthCheckHealthyCount(props?: cloudwatch.MetricOptions): cloudwatch.Metric {
    return this.metric("ChildHealthCheckHealthyCount", { statistic: cloudwatch.Statistic.AVERAGE, ...props });
  }

  public addChildHealthCheck(healthCheck: IHealthCheck) {
    this.childHealthChecks.push(healthCheck);
  }
}
