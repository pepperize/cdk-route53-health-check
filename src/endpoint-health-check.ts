import { Annotations, Tags } from "aws-cdk-lib";
import * as cloudwatch from "aws-cdk-lib/aws-cloudwatch";
import * as route53 from "aws-cdk-lib/aws-route53";
import { Construct } from "constructs";
import { HealthCheckBase, HealthCheckOptions } from "./health-check";
import { HealthCheckType } from "./health-check-type";

export interface EndpointHealthCheckProps extends HealthCheckOptions {
  /**
   * The domain name that Route53 performs health checks on. Route53 resolves the IP address and performs the lookup.
   *
   * If IP address is given, it's used as the host name.
   *
   * <b>Either DomainName or IpAddress must be specified</b>
   */
  readonly domainName?: string;
  /**
   * The ip address that Route53 performs health checks on. Optionally a domain name may be given.
   *
   * <b>An IP address must be specified if protocol TCP</b>
   */
  readonly ipAddress?: string;
  /**
   * The protocol that Route53 uses to communicate with the endpoint.
   *
   * <b>An IP address must be specified if protocol TCP</b>
   *
   * @default HTTPS
   */
  readonly protocol?: Protocol;
  /**
   * The port that Route53 performs health checks. The port must be between 1 and 65535.
   *
   * @default 80 for HTTP; 443 for HTTPS
   */
  readonly port?: number;
  /**
   * Specify that Route53 sends the host name for TLS negotiation.
   *
   * @default true for HTTPS
   */
  readonly enableSni?: boolean;
  /**
   * The path for HTTP or HTTPS health checks. Provide a string between 1 and 255 length.
   */
  readonly resourcePath?: string;
  /**
   * The search string for HTTP or HTTPS health checks. Route53 will search in the response body. Provide a string between 1 and 255 length.
   */
  readonly searchString?: string;

  /**
   * The number of seconds between the time that Route53 gets a response from your endpoint and the time that it sends the next health check request. Each Route53 health checker makes requests at this interval. Provide a number between 10 and 30.
   *
   * <i>If you choose an interval of 10 and there are 15 health checkers, the endpoint will receive approximately 1 request per second.</i>
   *
   * <b>Can't be changed after HealthCheck is deployed</b>
   */
  readonly requestInterval?: number;
  /**
   * The number of consecutive health checks that an endpoint must pass or fail for Route53 to change the current status of the endpoint between healthy and unhealthy. Provide a number between 1 and 10.
   */
  readonly failureThreshold?: number;
  /**
   * Whether Route53 measures the latency between health checkers in multiple AWS regions and your endpoint, and displays a CloudWatch latency graphs in the Route53 console.
   *
   * <b>Can't be changed after HealthCheck is deployed</b>
   */
  readonly latencyGraphs?: boolean;
  /**
   * The list of regions from which Route53 health checkers check the endpoint.
   *
   * <b>If omitted Route53 performs checks from all health checker regions.</b>
   */
  readonly regions?: HealthCheckerRegions[];
}

/**
 * Create a Route53 HealthCheck that monitors an endpoint either by domain name or by IP address.
 *
 * <b>Example</b>
 * ```typescript
 * new EndpointHealthCheck(stack, "HealthCheck", {
 *   domainName: "pepperize.com",
 * });
 * ```
 * Generates
 * ```yaml
 * Resources:
 *   Type: AWS::Route53::HealthCheck
 *   Properties:
 *     HealthCheckConfig:
 *       FullyQualifiedDomainName: "pepperize.com"
 *       Port: 443
 *       Type: "HTTPS"
 *       EnableSNI: true
 * ```
 * @link https://docs.aws.amazon.com/de_de/AWSCloudFormation/latest/UserGuide/aws-resource-route53-healthcheck.html#aws-resource-route53-healthcheck-properties
 *
 * @resource AWS::Route53::HealthCheck
 */
export class EndpointHealthCheck extends HealthCheckBase {
  public readonly healthCheckId: string;

  constructor(scope: Construct, id: string, props: EndpointHealthCheckProps) {
    super(scope, id);

    const protocol = props.protocol || Protocol.HTTPS;
    const type = this.healthCheckType(protocol, props.searchString);
    const port = this.defaultPort(props.port, type);
    const enableSni = this.enableSniForHttps(type, props.enableSni);

    if (!props.domainName && !props.ipAddress) {
      Annotations.of(this).addError("Either DomainName or IpAddress has to be specified");
    }

    if (props.ipAddress && !new RegExp(IP_ADDRESS_REGEX_PATTERN).test(props.ipAddress)) {
      Annotations.of(this).addError("IpAddress must be valid");
    }

    if (undefined != props.port && (props.port < 1 || props.port > 65535)) {
      Annotations.of(this).addError("Port has to be between 1 and 65535");
    }

    if (props.resourcePath && props.resourcePath.length > 255) {
      Annotations.of(this).addError("ResourcePath must be at least 255 character long");
    }

    if (props.searchString && props.searchString.length > 255) {
      Annotations.of(this).addError("SearchString must be at least 255 character long");
    }

    if (undefined != props.failureThreshold && (props.failureThreshold < 1 || props.failureThreshold > 10)) {
      Annotations.of(this).addError("FailureThreshold has to be between 10 and 30");
    }

    if (undefined != props.requestInterval && (props.requestInterval < 10 || props.requestInterval > 30)) {
      Annotations.of(this).addError("RequestInterval has to be between 10 and 30");
    }

    if (props.regions && props.regions.length < 3) {
      Annotations.of(this).addError("At least three HealthCheckerRegions have to be given");
    }

    const resource = new route53.CfnHealthCheck(this, "Resource", {
      healthCheckConfig: {
        enableSni: enableSni,
        fullyQualifiedDomainName: props.domainName,
        ipAddress: props.ipAddress,
        inverted: props.inverted,
        port: port,
        resourcePath: props.resourcePath,
        searchString: props.searchString,
        type: type,

        requestInterval: props.requestInterval,
        failureThreshold: props.failureThreshold,
        measureLatency: props.latencyGraphs,
        regions: props.regions,
      },
      healthCheckTags: this.tags.renderedTags,
    });

    this.healthCheckId = resource.attrHealthCheckId;

    const healthCheckName = props.healthCheckName || props.domainName;
    if (healthCheckName) {
      Tags.of(this).add("Name", healthCheckName);
    }
  }

  /**
   * Returns the CFN HealthCheckType for the given protocol. If undefined returns default HTTPS.
   */
  private healthCheckType(protocol?: Protocol, searchString?: string): HealthCheckType {
    switch (true) {
      case Protocol.TCP == protocol:
        return HealthCheckType.TCP;
      case Protocol.HTTPS == protocol && !!searchString:
        return HealthCheckType.HTTPS_STR_MATCH;
      case Protocol.HTTPS == protocol:
        return HealthCheckType.HTTPS;
      case Protocol.HTTP == protocol && !!searchString:
        return HealthCheckType.HTTP_STR_MATCH;
      case Protocol.HTTP == protocol:
        return HealthCheckType.HTTP;
      default:
        return HealthCheckType.HTTPS;
    }
  }

  /**
   * Sets the default if undefined for HTTP and HTTPS
   */
  private defaultPort(port?: number, type?: HealthCheckType): number | undefined {
    switch (true) {
      case port && port > 1:
        return port;
      case HealthCheckType.HTTPS == type:
        return 443;
      case HealthCheckType.HTTPS_STR_MATCH == type:
        return 443;
      case HealthCheckType.HTTP == type:
        return 80;
      case HealthCheckType.HTTP_STR_MATCH == type:
        return 80;
      default:
        return undefined;
    }
  }

  /**
   * Enables SNI by default for HTTPS if omitted, otherwise undefined
   */
  private enableSniForHttps(type?: HealthCheckType, enableSni?: boolean): boolean | undefined {
    if (HealthCheckType.HTTPS != type && HealthCheckType.HTTPS_STR_MATCH != type) {
      return undefined;
    }
    if (undefined == enableSni) {
      return true;
    }

    return enableSni;
  }

  /**
   * The percentage of Route53 health checkers that report that the status of the health check is healthy
   *
   * <b>LatencyGraphs has to be enabled</b>
   *
   * Valid statistics: Average (recommended), Minimum, Maximum
   */
  public metricHealthCheckPercentageHealthy(props?: cloudwatch.MetricOptions): cloudwatch.Metric {
    return this.metric("HealthCheckPercentageHealthy", { statistic: cloudwatch.Statistic.AVERAGE, ...props });
  }

  /**
   * The time in milliseconds that it took Route53 health checkers to establish a TCP connection with the endpoint
   *
   * Valid statistics: Average (recommended), Minimum, Maximum
   */
  public metricConnectionTime(props?: cloudwatch.MetricOptions): cloudwatch.Metric {
    return this.metric("ConnectionTime", { statistic: cloudwatch.Statistic.AVERAGE, ...props });
  }

  /**
   * The time in milliseconds that it took Route53 health checkers to complete the SSL/TLS handshake
   *
   * Valid statistics: Average, Minimum, Maximum
   */
  public metricSSLHandshakeTime(props?: cloudwatch.MetricOptions): cloudwatch.Metric {
    return this.metric("SSLHandshakeTime", { statistic: cloudwatch.Statistic.AVERAGE, ...props });
  }

  /**
   * The time in milliseconds that it took Route53 health checkers to receive the first byte of the response to an HTTP or HTTPS request
   *
   * Valid statistics: Average (recommended), Minimum, Maximum
   */
  public metricTimeToFirstByte(props?: cloudwatch.MetricOptions): cloudwatch.Metric {
    return this.metric("TimeToFirstByte", { statistic: cloudwatch.Statistic.AVERAGE, ...props });
  }
}

/**
 * The protocol that Route53 uses to communicate with the endpoint.
 */
export enum Protocol {
  HTTP = "HTTP",
  HTTPS = "HTTPS",
  TCP = "TCP",
}

/**
 * The regions of health checker from which Route53 performs checks on the endpoint.
 */
export enum HealthCheckerRegions {
  US_EAST_1 = "us-east-1",
  US_WEST_1 = "us-west-1",
  US_WEST_2 = "us-west-2",
  EU_WEST_1 = "eu-west-1",
  AP_SOUTHEAST_1 = "ap-southeast-1",
  AP_SOUTHEAST_2 = "ap-southeast-2",
  AP_NORTHEAST_1 = "ap-northeast-1",
  SA_EAST_1 = "sa-east-1",
}

const IP_ADDRESS_REGEX_PATTERN =
  "(^((([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5]))$|^(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9]))$)";
