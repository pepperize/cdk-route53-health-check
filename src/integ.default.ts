import { App, Stack } from "aws-cdk-lib";
import * as apigateway from "aws-cdk-lib/aws-apigateway";
import * as acm from "aws-cdk-lib/aws-certificatemanager";
import * as cloudfront from "aws-cdk-lib/aws-cloudfront";
import * as route53 from "aws-cdk-lib/aws-route53";
import * as targets from "aws-cdk-lib/aws-route53-targets";
import { EndpointHealthCheck } from "./endpoint-health-check";

const app = new App();

const shared = new Stack(app, "Shared");
const hostedZone = new route53.PublicHostedZone(shared, "HostedZone", {
  zoneName: "pepperize.com",
});

const primary = new Stack(app, "Primary");

const distribution = cloudfront.Distribution.fromDistributionAttributes(primary, "Distribution", {
  domainName: "pepperize.com",
  distributionId: "d111111abcdef8",
});
const recordSetPrimary = new route53.ARecord(primary, "RecordSetPrimary", {
  recordName: "www.pepperize.com",
  zone: hostedZone,
  target: route53.RecordTarget.fromAlias(new targets.CloudFrontTarget(distribution)),
});

const healthCheckPrimary = new EndpointHealthCheck(primary, "HealthCheckPrimary", {
  domainName: "www.pepperize.com",
});
healthCheckPrimary.failoverPrimary(recordSetPrimary);

const secondary = new Stack(app, "Secondary");

const certificate = acm.Certificate.fromCertificateArn(
  secondary,
  "Certificate",
  "arn:aws:acm:region:account:certificate/certificate_ID"
);
const restApi = new apigateway.RestApi(secondary, "RestApi", {
  domainName: { domainName: "www-1.pepperize.com", certificate: certificate },
});
restApi.root.addResource("{proxy+}", {}).addMethod("ANY");
const recordSetSecondary = new route53.ARecord(secondary, "RecordSetSecondary", {
  recordName: "www-1.pepperize.com",
  zone: hostedZone,
  target: route53.RecordTarget.fromAlias(new targets.ApiGateway(restApi)),
});
const healthCheckSecondary = new EndpointHealthCheck(secondary, "HealthCheckSecondary", {
  domainName: "www-1.pepperize.com",
});
healthCheckSecondary.failoverSecondary(recordSetSecondary, true);

app.synth();
