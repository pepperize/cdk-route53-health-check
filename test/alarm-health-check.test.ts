import { Stack } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import * as cloudwatch from "aws-cdk-lib/aws-cloudwatch";
import { AlarmHealthCheck } from "../src";

describe("AlarmHealthCheck", () => {
  it("Should match snapshot", () => {
    // Given
    const stack = new Stack(undefined, undefined, { env: { account: "123456789012", region: "us-east-1" } });
    const alarm = cloudwatch.Alarm.fromAlarmArn(
      stack,
      "Alarm",
      "arn:aws:cloudwatch:us-east-1:123456789012:alarm:any-alarm"
    );
    new AlarmHealthCheck(stack, "HealthCheck", {
      alarm: alarm,
    });

    // When
    const template = Template.fromStack(stack);

    // Then
    expect(template).toMatchSnapshot();
  });
  it("Should have resource", () => {
    // Given
    const stack = new Stack(undefined, undefined, { env: { account: "123456789012", region: "us-east-1" } });
    const alarm = cloudwatch.Alarm.fromAlarmArn(
      stack,
      "Alarm",
      "arn:aws:cloudwatch:us-east-1:123456789012:alarm:any-alarm"
    );
    new AlarmHealthCheck(stack, "HealthCheck", {
      healthCheckName: "alarm",
      alarm: alarm,
    });

    // When
    const template = Template.fromStack(stack);

    // Then
    template.hasResourceProperties("AWS::Route53::HealthCheck", {
      HealthCheckConfig: {
        AlarmIdentifier: {
          Name: "any-alarm",
          Region: "us-east-1",
        },
      },
      HealthCheckTags: [
        {
          Key: "Name",
          Value: "alarm",
        },
      ],
    });
  });
});
