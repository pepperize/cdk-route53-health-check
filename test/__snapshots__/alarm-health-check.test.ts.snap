// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlarmHealthCheck Should match snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "HealthCheckA1C381C7": Object {
      "Properties": Object {
        "HealthCheckConfig": Object {
          "AlarmIdentifier": Object {
            "Name": "any-alarm",
            "Region": "us-east-1",
          },
          "InsufficientDataHealthStatus": "LastKnownStatus",
          "Type": "CLOUDWATCH_METRIC",
        },
      },
      "Type": "AWS::Route53::HealthCheck",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
